libraries/AP_Logger/AP_Logger.cpp

```diff
@@ -1394,10 +1399,85 @@ bool AP_Logger::log_while_disarmed(void) const
 }
 
 #if HAL_LOGGER_FILE_CONTENTS_ENABLED
+void AP_Logger::prepare_at_arming_sys_file_logging()
+{
+    // free existing content:
+    at_arm_file_content.reset();
+
+    /*
+      log files useful for diagnostics on arming. We log on arming as
+      with LOG_DISARMED we don't want to log the statistics at boot or
+      we wouldn't get a realistic idea of key system values
+      Note that some of these files may not exist, in that case they
+      are ignored
+     */
+    static const char *log_content_filenames[] = {
+        "@SYS/uarts.txt",
+        "@SYS/dma.txt",
+        "@SYS/memory.txt",
+        "@SYS/threads.txt",
+        "@ROMFS/hwdef.dat",
+        "@SYS/storage.bin",
+        "@SYS/crash_dump.bin",
+    };
+    for (const auto *name : log_content_filenames) {
+        log_file_content(at_arm_file_content, name);
+    }
+}
+
+void AP_Logger::FileContent::reset()
+{
+    WITH_SEMAPHORE(sem);
+    file_list *next = nullptr;
+    for (auto *c = head; c != nullptr; c = next) {
+        next = c->next;
+        delete [] c->filename;
+        delete c;
+    }
+    head = nullptr;
+    tail = nullptr;
+    if (fd != -1) {
+        AP::FS().close(fd);
+        fd = -1;
+    }
+    counter = 0;
+    fast = false;
+    offset = 0;
+}
+
+// removes victim from FileContent ***and delete()s it***
+void AP_Logger::FileContent::remove_and_free(file_list *victim)
+{
+    WITH_SEMAPHORE(sem);
+
+    file_list *prev = nullptr;
+    for (auto *c = head; c != nullptr; prev = c, c = c->next) {
+        if (c != victim) {
+            continue;
+        }
+
+        // found the item to remove; remove it and return
+        if (prev == nullptr) {
+            head = victim->next;
+        } else {
+            prev->next = victim->next;
+        }
+        delete [] victim->filename;
+        delete victim;
```

Comment:
do you need an FD close() here?

Reply:
The fd is a property of the file-content-transfer mechanism rather than any particular entry in the file list.

You can end up in the strange situation where you've removed from the list the entry for the fd - but I believe that's probably OK....